pallets/xyk/src/lib.rs

```diff
@@ -731,27 +734,25 @@ impl<T: Config> AMM<T::AccountId, AssetId, AssetPair, Balance> for Pallet<T> {
 			);
 		}
 
+		let buy_price = hydra_dx_math::calculate_in_given_out(asset_out_reserve, asset_in_reserve, amount)
+			.map_err(|_| Error::<T>::BuyAssetAmountInvalid)?;
+
 		let transfer_fee = if discount {
-			Self::calculate_discounted_fee(amount)?
+			Self::calculate_discounted_fee(buy_price)?
 		} else {
-			Self::calculate_fee(amount)?
+			Self::calculate_fee(buy_price)?
 		};
 
-		ensure!(
-			amount + transfer_fee <= asset_out_reserve,
-			Error::<T>::InsufficientPoolAssetBalance
-		);
-
-		let buy_price =
-			hydra_dx_math::calculate_in_given_out(asset_out_reserve, asset_in_reserve, amount + transfer_fee)
-				.map_err(|_| Error::<T>::BuyAssetAmountInvalid)?;
+		let buy_price_with_fee = buy_price
+			.checked_add(transfer_fee)
+			.ok_or(Error::<T>::BuyAssetAmountInvalid)?;
 
 		ensure!(
-			T::Currency::free_balance(assets.asset_in, who) >= buy_price,
+			T::Currency::free_balance(assets.asset_in, who) >= buy_price_with_fee,
 			Error::<T>::InsufficientAssetBalance
 		);
 
-		ensure!(max_limit >= buy_price, Error::<T>::AssetBalanceLimitExceeded);
+		ensure!(max_limit >= buy_price_with_fee, Error::<T>::AssetBalanceLimitExceeded);
```

Comment:
Perhaps you can move this check before the free balance check.

Reply:
done